### DISCUSSION TITLE
使用Use AdGuard browsing security web service功能导致DNS解析异常

### DISCUSSION THREAD

**Discussion by WindyCloudCute** [AdguardTeam/AdGuardHome#3714] [Q&A]
### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [ √] I am running the latest version
- [ √] I checked the documentation and found no answer
- [ √] I checked to make sure that this issue has not already been filed

### Issue Details

> AdGuardHome -v --version
> AdGuard Home
> Version: v0.106.3
> Channel: release
> Go version: go1.15.11
> Build time: 2021-05-19T13:25:09Z+0000
> GOOS: linux
> GOARCH: amd64
> Race: false
> Main module:
>         github.com/AdguardTeam/AdGuardHome (devel)
> Dependencies：

>         github.com/AdguardTeam/dnsproxy@v0.37.4 (sum: h1:YIoJkIp828LKmmmgxXvZHUKfGLsqTQAK8g+4DXbDbyU=)
>         github.com/AdguardTeam/golibs@v0.5.0 (sum: h1:qwhEKjDrT0UcwDnHrNU2Yg/DLR9b/GsUncnXYW6VzAU=)
>         github.com/AdguardTeam/urlfilter@v0.14.5 (sum: h1:WyF0hg0MwKevsqNPkoaZFH8f5WRi/yuy/7qePtYt5Ts=)
>         github.com/NYTimes/gziphandler@v1.1.1 (sum: h1:ZUDjpQae29j0ryrS0u/B8HZfJBtBQHjqw2rQ2cqUQ3I=)
>         github.com/aead/chacha20@v0.0.0-20180709150244-8b13a72661da (sum: h1:KjTM2ks9d14ZYCvmHS9iAKVt9AyzRSqNU1qabPih5BY=)
>         github.com/aead/poly1305@v0.0.0-20180717145839-3fee0db0b635 (sum: h1:52m0LGchQBBVqJRyYYufQuIbVqRawmubW3OFGqK1ekw=)
>         github.com/ameshkov/dnscrypt/v2@v2.1.3 (sum: h1:DG4Uf7LSDg6XDj9sp3maxh3Ur26jeGQaP5MeYosn6v0=)
>         github.com/ameshkov/dnsstamps@v1.0.3 (sum: h1:Srzik+J9mivH1alRACTbys2xOxs0lRH9qnTA7Y1OYVo=)
>         github.com/beefsack/go-rate@v0.0.0-20200827232406-6cde80facd47 (sum: h1:M57m0xQqZIhx7CEJgeLSvRFKEK1RjzRuIXiA3HfYU7g=)
>         github.com/cheekybits/genny@v1.0.0 (sum: h1:uGGa4nei+j20rOSeDeP5Of12XVm7TGUd4dJA9RDitfE=)
>         github.com/digineo/go-ipset/v2@v2.2.1 (sum: h1:k6skY+0fMqeUjjeWO/m5OuWPSZUAn7AucHMnQ1MX77g=)
>         github.com/fsnotify/fsnotify@v1.4.9 (sum: h1:hsms1Qyu0jgnwNXIxa+/V/PDsU6CfLf6CNO8H7IWoS4=)
>         github.com/go-ping/ping@v0.0.0-20210216210419-25d1413fb7bb (sum: h1:2opwLSXqxE0Za64PdpskXuvLYDj/XHQAD8tLcYpSlvY=)
>         github.com/gobuffalo/envy@v1.9.0 (sum: h1:eZR0DuEgVLfeIb1zIKt3bT4YovIMf9O9LXQeCZLXpqE=)
>         github.com/gobuffalo/packd@v1.0.0 (sum: h1:6ERZvJHfe24rfFmA9OaoKBdC7+c9sydrytMg8SdFGBM=)
>         github.com/gobuffalo/packr@v1.30.1 (sum: h1:hu1fuVR3fXEZR7rXNW3h8rqSML8EVAf6KNm0NKO/wKg=)
>         github.com/google/renameio@v1.0.1-0.20210406141108-81588dbe0453 (sum: h1:vvZyWjAX9oDB+DWpMsZMyv6Q3NZtim2C5Zcdh+H0OmQ=)
>         github.com/AdguardTeam/dhcp@v0.0.0-20210517101438-550ef4cd8c6e (sum: h1:M6YnFP12o0/SjBEPt6b2r8ZkIy/wsV14TK8X9Tb6DEE=)
>         github.com/joho/godotenv@v1.3.0 (sum: h1:Zjp+RcGpHhGlrMbJzXTrZZPrWj+1vfm90La1wgB6Bhc=)
>         github.com/joomcode/errorx@v1.0.3 (sum: h1:3e1mi0u7/HTPNdg6d6DYyKGBhA5l9XpsfuVE29NxnWw=)
>         github.com/josharian/native@v0.0.0-20200817173448-b6b71def0850 (sum: h1:uhL5Gw7BINiiPAo24A2sxkcDI0Jt/sqp1v5xQCniEFA=)
>         github.com/kardianos/service@v1.2.0 (sum: h1:bGuZ/epo3vrt8IPC7mnKQolqFeYJb7Cs8Rk4PSOBB/g=)
>         github.com/lucas-clemente/quic-go@v0.20.1 (sum: h1:hb5m76V8QS/8Nw/suHvXqo3BMHAozvIkcnzpJdpanSk=)
>         github.com/marten-seemann/qtls-go1-15@v0.1.4 (sum: h1:RehYMOyRW8hPVEja1KBVsFVNSm35Jj9Mvs5yNoZZ28A=)
>         github.com/mdlayher/ethernet@v0.0.0-20190606142754-0394541c37b7 (sum: h1:lez6TS6aAau+8wXUP3G9I3TGlmPFEq2CTxBaRqY6AGE=)
>         github.com/mdlayher/netlink@v1.4.0 (sum: h1:n3ARR+Fm0dDv37dj5wSWZXDKcy+U0zwcXS3zKMnSiT0=)
>         github.com/mdlayher/raw@v0.0.0-20191009151244-50f2db8cc065 (sum: h1:aFkJ6lx4FPip+S+Uw4aTegFMct9shDvP+79PsSxpm3w=)
>         github.com/miekg/dns@v1.1.40 (sum: h1:pyyPFfGMnciYUk/mXpKkVmeMQjfXqt3FAJ2hy7tPiLA=)
>         github.com/patrickmn/go-cache@v2.1.0+incompatible (sum: h1:HRMgzkcYKYpi3C8ajMPV8OFXaaRUnok+kx1WdO15EQc=)
>         github.com/pkg/errors@v0.9.1 (sum: h1:FEBLx1zS214owpjy7qsBeixbURkuhQAwrK5UwLGTwt4=)
>         github.com/rogpeppe/go-internal@v1.7.0 (sum: h1:3qqXGV8nn7GJT65debw77Dzrx9sfWYgP0DDo7xcMFRk=)
>         github.com/satori/go.uuid@v1.2.0 (sum: h1:0uYX9dsZ2yD7q2RtLRtPSdGDWzjeM3TbMJP9utgA0ww=)
>         github.com/ti-mo/netfilter@v0.4.0 (sum: h1:rTN1nBYULDmMfDeBHZpKuNKX/bWEXQUhe02a/10orzg=)
>         github.com/u-root/u-root@v7.0.0+incompatible (sum: h1:u+KSS04pSxJGI5E7WE4Bs9+Zd75QjFv+REkjy/aoAc8=)
>         go.etcd.io/bbolt@v1.3.5 (sum: h1:XAzx9gjCb0Rxj7EoqcClPD1d5ZBxZJk0jbuoPHenBt0=)
>         golang.org/x/crypto@v0.0.0-20210220033148-5ea612d1eb83 (sum: h1:/ZScEX8SfEmUGRHs0gxpqteO5nfNW6axyZbBdw9A12g=)
>         golang.org/x/net@v0.0.0-20210405180319-a5a99cb37ef4 (sum: h1:4nGaVu0QrbjT/AK2PRLuQfQuh6DJve+pELhqTdAj3x0=)
>         golang.org/x/sys@v0.0.0-20210330210617-4fbd30eecc44 (sum: h1:Bli41pIlzTzf3KEY06n+xnzK/BESIg2ze4Pgfh/aI8c=)
>         golang.org/x/text@v0.3.5 (sum: h1:i6eZZ+zk0SOf0xgBpEpPD18qWcJda6q1sxt3S0kzyUQ=)
>         gopkg.in/natefinch/lumberjack.v2@v2.0.0 (sum: h1:1Lc07Kr7qY4U2YPouBjpCLxpiyxIVoxqXgkXLknAOE8=)
>         gopkg.in/yaml.v2@v2.4.0 (sum: h1:D8xgwECY7CYvx+Y2n4sBz93Jn9JRvxdiyyo8CTfuKaY=)
>         howett.net/plist@v0.0.0-20201203080718-1454fab16a06 (sum: h1:QDxUo/w2COstK1wIBYpzQlHX/NqaQTcf9jyz347nI58=)

* **Version of AdGuard Home server:**
v0.106.3
* **How did you install AdGuard Home:**
GitHub releases
* **How did you setup DNS configuration:**
Router
* **If it's a router or IoT, please write device model:**
Intel
* **CPU architecture:**
AMD64
* **Operating system and version:**
linux ,Windows10

### Expected Behavior
需要启用Use AdGuard browsing security web service

### Actual Behavior
在启用Use AdGuard browsing security web service功能后，DNS出现无法解析，网页无法访问但IP连接不会被阻断。

**Comment by Hill-98** [comment]
这个问题是由于局域网的全局代理造成的，因为代理软件也需要解析域名，但是代理软件解析时 AdGraud 也会检查，但是这时由于检查请求被转发到代理软件，所以无法请求成功，就会一直死循环，可通过 AdGraud 的 DNS 重写将代理软件用到的域名直接返回解析结果。

**Comment by ameshkov** [comment]
我不懂解析怎么异常？你看，AdGuard browsing security内它其实用 AdGuard DNS 服务器检查域名。如果 AdGuard DNS 被拦截，无法用 browsing security。

**Reply by Hill-98** [reply]
> 我不懂解析怎么异常？你看，AdGuard browsing security内它其实用 AdGuard DNS 服务器检查域名。如果 AdGuard DNS 被拦截，无法用 browsing security。

This problem is common in China, where people often need proxies to access IP addresses around the world because of GFW.

So many people will set up transparent proxies in their routers via iptables to forward requests to proxy software (like V2ray), but the proxy software also needs to resolve the domain name, so it creates a dead end.

**Reply by ameshkov** [reply]
Ah, makes sense indeed.

**Comment by Hill-98** [comment]
The best practice is to use the DNS rewrite feature to assign an IP to the proxyer's domain name, which will jump out of the dead loop.

**Reply by WindyCloudCute** [reply]
> @ameshkov I would suggest introducing this question in the FAQ, which is probably common among Chinese users.

请问有没有办法排查这些问题，我没有办法查明是什么原因导致的死循环，事实上我代理是完全关闭的，也没有办法解决这个问题。

**Reply by ameshkov** [reply]
你能使AdGuard写verbose水平日志，然后检查日志文件，它是否有死循环。

https://github.com/AdguardTeam/AdGuardHome/wiki/FAQ#verboselog

**Reply by Hill-98** [reply]
没有代理的情况下在中国是无法访问 dns-family.adguard.com 的，所以必须使用代理，且不能让解析陷入死循环。

**Reply by WindyCloudCute** [reply]
> 你能使AdGuard写verbose水平日志，然后检查日志文件，它是否有死循环。
> 
> https://github.com/AdguardTeam/AdGuardHome/wiki/FAQ#verboselog

这太难了  : (

**Reply by WindyCloudCute** [reply]
我发现了一个办法，在代理软件里设置强代理adguard.com及其所有子域名，然后在AdGuard里设置自定义规则，`@@||adguard.com^`貌似就可以让功能发挥正常。

**Reply by Hill-98** [reply]
本质上还是让 dns-family.adguard.com 走代理，可能是因为你用的是 gfwlist 模式。

**Reply by Hill-98** [reply]
@ameshkov  Speaking of gfwlist, I'd also like to add to the FAQ solution that if a user is using gfwlist mode (a mode that only proxies domains that are blocked by GFW), be sure to force the proxy to dns-family.adguard.com.

**Reply by WindyCloudCute** [reply]
嗯我用的是绕过chinalist，但是我没有去看list里面是否包含adguard的域名，不过这个也可以作为一个社区面对中国用户的经验吧，毕竟我觉得出现这个问题的人应该不会很少，而且几乎搜索不到任何相关的资料。

**Comment by Hill-98** [comment]
@ameshkov I would suggest introducing this question in the FAQ, which is probably common among Chinese users.

**Reply by ameshkov** [reply]
What kind of setup would you suggest to add to FAQ?

**Reply by Hill-98** [reply]
Suggested steps:

```
Turn off secure browsing related services and check if you can access https://dns-family.adguard.com.

If it doesn't work, your firewall or carrier may be blocking access to it.
If so, you should be able to use it normally, or you can continue with the following checks.(emphasis on Chinese users)

If your LAN deploys proxy software and forwards some of the requests to the proxy software through your router,
it may be normal for the security service to not work. because the proxy software you are using may also need to
do DNS resolution, but before the proxy software can resolve successfully, AdGuard needs access to the security
service and that access request is forwarded to the proxy software, and you are stuck in a dead end.

Solution.
1. Use AdGuard's DNS rewrite to assign the proxy software's domain name to the corresponding IP
2. Proxy software uses IP for connection
```

My English is not very good and there may be areas that I need to improve.

**Reply by ameshkov** [reply]
Thank you! I've created a task about that: https://github.com/AdguardTeam/AdGuardHome/issues/3716
